src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -117,13 +127,33 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
   def findCampaignsByUpdate(update: UpdateId): Future[Seq[Campaign]] =
     db.run(campaignRepo.findByUpdateAction(update))
 
-  def campaignStats(campaignId: CampaignId): Future[CampaignStats] = for {
-    status    <- findClientCampaign(campaignId).map(_.status)
-    finished  <- countFinished(campaignId)
-    cancelled <- countCancelled(campaignId)
-    failed    <- failedDevices(campaignId)
-    stats     <- campaignStatsFor(campaignId)
-  } yield CampaignStats(campaignId, status, finished, failed, cancelled, stats)
+  /**
+   * Calculates campaign-wide statistic counters, also taking retry campaings
+   * into account if any exist.
+   */
+  def campaignStats(campaignId: CampaignId): Future[CampaignStats] = db.run {
+    val statsAction = for {
+      mainCampaign <- campaignRepo.findAction(campaignId)
+      retryCampaignIds <- campaignRepo.findChildIdsOfAction(campaignId)
+      (mainProcessed, mainAffected) <- countProcessedAndAffectedDevicesForAllOfAction(Set(campaignId))
+      (retryProcessed, retryAffected) <- countProcessedAndAffectedDevicesForAllOfAction(retryCampaignIds)
```

Comment:
Why is group stats is not counting cancelled?

Reply:
I don't know for certain, but it seems to me that counting cancelled devices using group stats is incorrect because of the ability to cancel individual devices. AFAIK it won't make it to the group stats.